erts/emulator/internal_doc/BeamAsm.md

```diff
@@ -0,0 +1,373 @@
+the same information, but it is easier to share with others as it does
+not need the symbols in the executable.
+
+Using the same data we can also produce a graph where the scheduler profile data
+has been merged by using `sed`:
+
+    ## Strip [0-9]+_ from all scheduler names
+    sed -e 's/^[0-9]\+_//' out.folded > out.folded_sched
+    ## Create the svg
+    flamegraph.pl out.folded_sched > out_sched.svg
+
+![Linux Perf FlameGraph: dialyzer PLT build](figures/perf-beamasm-merged.svg)
+
+You can view a larger version [here](seefile/figures/perf-beamasm-merged.svg).
+There are many different transformations that you can do to make the graph show
+you what you want.
+
+### Annotate perf functions
+
+If you want to be able to use the `perf annotate` functionality (and in extention
+the annotate functionality in the `perf report` gui) you need to use a monotonic
+clock when calling `perf record`, i.e. `perf record -k mono`. So for a dialyzer
+run you would do this:
+
+    ERL_FLAGS="+JPperf true +S 1" perf record -k mono --call-graph lbr \
+      dialyzer --build_plt -Wunknown --apps compiler crypto erts kernel stdlib \
+      syntax_tools asn1 edoc et ftp inets mnesia observer public_key \
+      sasl runtime_tools snmp ssl tftp wx xmerl tools
+
+In order to use the `perf.data` produced by this record you need to first call
+`perf inject --jit` like this:
+
+    perf inject --jit -i perf.data -o perf.jitted.data
+
+and then you can view an annotated function like this:
+
+    perf annotate -M intel -i perf.jitted.data erl_types:t_has_var/1
+
+or by pressing `a` in the `perf report` ui. Then you get something like this:
+
+![Linux Perf FlameGraph: dialyzer PLT build](figures/beamasm-perf-annotate.png)
+
+> *WARNING*: Calling `perf inject --jit` will create a lot of files in `/tmp/`
+> and in `~/.debug`. So make sure to cleanup in those directories from time to
+> time or you may run out of inodes.
+
+### perf tips and tricks
+
+You can do a lot of neat things with `perf`. Below is a list of some of the options
+we have found useful:
+
+* `perf report --no-children`
+    Do not include the accumulation of all children in a call.
+* `perf report  --call-graph callee`
+    Show the callee rather than the caller when expanding a function call.
+* `perf archive`
+    Create an archive with all the artifacts needed to inspect the data
+    on another host. In early version of perf this command does not work,
+    instead you can use [this bash script](https://github.com/torvalds/linux/blob/master/tools/perf/perf-archive.sh).
+
+## FAQ
+
+### Is the interpreter still available?
+
+Yes, you can still build the interpreter if you want to. In fact, it is what is used
+on platforms where BeamAsm does not yet work. You can either completely disable
+BeamAsm by passing `--disable-beamasm` to configure. Or you can build the
+interpreter using `make FLAVOR=emu` and then run it using `erl -emu_flavor emu`.
```

Comment:
Is it possible to build multiple flavors at once and pick which to use each boot time? Something that would make it easy to benchmark and evaluate perf changes.

Reply:
yes, it is possible. If you compile like above, you can run either `erl -emu_flavor asm` or `erl -emu_flavor emu`. `make release` and `make install` also work with `FLAVOR`, so you can install both `emu` and `asm` into your Erlang release.

Reply:
The latest release had changed `asm` to `jit` in the example above.